src/nodes/envelopes.rs

```diff
@@ -0,0 +1,452 @@
+//! AHDSR (attack, hold, decay, sustain, release) volume envelope.
+
+use bevy::ecs::component::Component;
+use firewheel::{
+    StreamInfo,
+    channel_config::{ChannelConfig, ChannelCount},
+    clock::{ClockSamples, ClockSeconds},
+    diff::{Diff, Patch},
+    dsp::volume::Volume,
+    event::NodeEventList,
+    node::{
+        AudioNode, AudioNodeInfo, AudioNodeProcessor, ConstructProcessorContext, EmptyConfig,
+        ProcBuffers, ProcInfo, ProcessStatus,
+    },
+    param::smoother::SmoothedParam,
+};
+use std::{num::NonZeroU32, ops::Range};
+
+use crate::timeline::{DiscreteTimeline, TimelineEvent};
+
+/// How an envelope responds to receiving [`TriggerState::On`] if it is not at rest.
+#[derive(Diff, Patch, Copy, Clone, PartialEq, Eq, Hash, Debug, Default)]
+pub enum RetriggerMode {
+    /// Restart the cycle of the envelope from the `attack` section, but use the current value
+    /// as the starting point.
+    #[default]
+    Normal,
+    /// Restart the envelope from `lo`.
+    Restart,
+    /// Ignore any new [`TriggerState::On`] events unless the envelope is at rest.
+    Ignore,
+}
+
+impl RetriggerMode {
+    fn state_transition(&self) -> fn(f32, Option<AhdsrState>) -> (f32, AhdsrState) {
+        match self {
+            RetriggerMode::Ignore => |cur, state| (cur, state.unwrap_or(AhdsrState::Attack)),
+            RetriggerMode::Normal => |cur, _| (cur, AhdsrState::Attack),
+            RetriggerMode::Restart => |_, _| (0., AhdsrState::Attack),
+        }
+    }
```

Comment:
This seems overwrought. Are you expecting better performance with this approach? Is it really faster to jump to a function pointer than to check an enum with three variants?

I'd prefer this to be a simple method on `RetriggerMode`. Even if this performs better, I doubt it's by enough of a margin to warrant potentially tricker long-term maintenance.